iAPI Docs: Add client-side navigation guide under "Core Concepts" (#75263)
WordPress/gutenberg: docs/reference-guides/interactivity-api/core-concepts/client-side-navigation.md

# Client-Side Navigation

Client-side navigation is a technique that allows navigation between pages without requiring a full page reload. Instead of the browser fetching an entirely new HTML document from the server, client-side navigation fetches the new page's content and updates only the parts of the DOM that have changed. This results in faster, smoother page transitions and a more app-like user experience.

The Interactivity API provides client-side navigation through the `@wordpress/interactivity-router` package. The central concept is the **router region**: a section of your page that the router knows how to update during navigation. You mark these sections with the `data-wp-router-region` directive, and when the user navigates to a new URL, the router fetches the destination page and replaces only the content inside matching regions — leaving everything else on the page untouched.

> This is known as **region-based client-side navigation**, and it is the recommended approach for implementing client-side navigation in WordPress. There is also an experimental **full-page client-side navigation** mode, which treats the entire `<body>` element as a single region — effectively updating the whole page content without a traditional reload. Full-page navigation is covered at the end of this guide in [Full-page client-side navigation (experimental)](#full-page-client-side-navigation-experimental).

Reviewer: Let's do this using bullet points, it's going to be clearer.

```suggestion
The Interactivity API supports two navigation modes:

- **Region-based client-side navigation** — The recommended approach for implementing client-side navigation in WordPress.
- **Full-page client-side navigation** _(experimental)_ — Treats the entire `<body>` element as a single region, effectively updating the whole page content without a traditional reload. Covered at the end of this guide in [Full-page client-side navigation (experimental)](#full-page-client-side-navigation-experimental).
```

Author: Done in https://github.com/WordPress/gutenberg/pull/75263/commits/87ec5b0cd40199ebcbd32a43db4ce541ff44df44

## How client-side navigation works

> When a user triggers a navigation (for example, by clicking a link), the Interactivity Router:

Reviewer: Let's make it clearer that it's not just a link, but a link that calls `actions.navigate()`.

```suggestion
When a user triggers a navigation, for example, by clicking a link that has a `data-wp-on--click` directive that calls `actions.navigate()`, the Interactivity Router:
```

Author: Clarified in https://github.com/WordPress/gutenberg/pull/75263/commits/c2da85a637055190e2ce22e6a8f4cd716d5b7cf7.

6. **Handles accessibility**: Screen reader announcements are made to indicate navigation progress.

This approach offers several benefits:

- **Improved performance**: Only the changed parts of the page are updated, reducing data transfer and DOM manipulation.
- **Preserved state**: Client-side state (global state, local context) is preserved across navigations.
- **Smooth transitions**: No flash of white screen between pages; transitions feel instant and app-like.
- **SEO-friendly**: Since the server still renders complete HTML pages, search engines can crawl your site normally.

## Getting started with the Interactivity Router

The `@wordpress/interactivity-router` package is bundled with WordPress Core since version 6.5. If you are starting a new project, the easiest way to get set up is using the [`@wordpress/create-block-interactive-template`](https://www.npmjs.com/package/@wordpress/create-block-interactive-template) scaffolding tool, which creates a block with the Interactivity API already configured:

```bash
npx @wordpress/create-block@latest my-interactive-block --template @wordpress/create-block-interactive-template

> ```

Reviewer: @juanmaguitar, weren't you working on a template for client-side navigation? Could it be?

Reviewer 2: I think you may refer to the example I was working on for the block development examples:
https://github.com/WordPress/block-development-examples/tree/trunk/plugins/interactivity-router-2f43f8

But it would be great to have a template for client-side-navigation. In fact, we could expand the existing `@wordpress/create-block-interactive-template` with a new variant for clien-side navigation (now it's easier because each variant can have its own structure of files https://github.com/WordPress/gutenberg/issues/67768)

Reviewer: Nice. Thanks Juanma 🙂

If you already have an interactive block and want to add client-side navigation, you need to:

1. **Add the router dependency**: Add `@wordpress/interactivity-router` as a dependency of your block's script module. This is typically done by dynamically importing the package in your `view.js` file (as shown in the examples below), which ensures it is only loaded when needed.
2. **Define router regions**: Mark the HTML elements that should be updated during navigation using the `data-wp-router-region` attribute.
3. **Trigger navigation**: Use the router's `actions.navigate()` function to navigate programmatically when the user interacts with your block.

> For detailed API documentation, see the [`@wordpress/interactivity-router` package README](/packages/interactivity-router/README.md).

Reviewer: There is really nothing in the README that is not explained here, so I would try to do the opposite: I would make this document the "source of truth" and redirect users here, instead of the other way around.

So I would delete this comment.

```suggestion
```

Author: Done in https://github.com/WordPress/gutenberg/pull/75263/commits/e06e9ca2d204a0d4a45f74abfc34929b63da7e20.

Here's how to implement a link that navigates client-side. First, the HTML in your block's `render.php` connects the link's click event to the `navigateTo` action:

```html
<a
    data-wp-on--click="actions.navigateTo"
    href="/page-2/"
>
    Go to Page 2
</a>
```

<div class="callout callout-info">
This element must be placed inside an element with the <code>data-wp-interactive="myPlugin"</code> directive (like the router region defined above), so the directive knows which store namespace to look up the action in. Alternatively, you can specify the namespace explicitly in the directive value itself: <code>data-wp-on--click="myPlugin::actions.navigateTo"</code>. For more details on how namespaces work, see the <a href="https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/directives-and-store/">Interactivity API Reference</a>.

> </div>

Reviewer: I don't think this is relevant here and it could even be a bit confusing because it seems like these directives have some kind of special requirement regarding the namespace, when they are normal directives. I would remove it.

```suggestion
```

Author: Done in https://github.com/WordPress/gutenberg/pull/75263/commits/7309d5db4dc7564cd407f61d0487d4b6a49149fb

- The block is **not interactive** (it renders only static HTML), or
- The block **is interactive and uses the Interactivity API** for all its client-side behavior.

Set it to `false` (or leave it unset) if the block is interactive but uses vanilla JavaScript, jQuery, or any other JavaScript framework or library other than the Interactivity API.

### Non-interactive blocks

Blocks that render only static HTML — with no JavaScript-driven interactivity — are inherently compatible with client-side navigation. Since they have no client-side state to manage or event listeners to re-attach, they can be safely re-rendered by the router at any time.

For these blocks, set only `clientNavigation` to `true`:

```json
{
    "supports": {
        "interactivity": {
            "clientNavigation": true
        }
    }
}
```

Examples of non-interactive blocks include the Paragraph, Heading, Image, and List blocks. Even though they don't use Interactivity API directives, they are fully compatible with client-side navigation because their output is purely static HTML.

### Interactive blocks using the Interactivity API

> Blocks that use the Interactivity API for their client-side behavior are compatible with client-side navigation. The Interactivity API's directive system integrates with the router, so interactive state, event handlers, and reactive bindings are properly preserved or re-initialized when regions are updated during navigation.

Reviewer: This is incorrect; some blocks, even using the Interactivity API, are not compatible with client-side navigation. This is the part we need to work on now to fully understand all the requirements that are needed.

For now, I would delete this entire section and work on it in a subsequent PR, as I think it's going to take us some time.

Author: I've removed the section in https://github.com/WordPress/gutenberg/pull/75263/commits/fde55bad5982ed63e29559db948d9e145e296fa1.

For these blocks, set both properties to `true` (or use the shorthand):

```json
{
    "supports": {
        "interactivity": true
    }
}
```

### Interactive blocks using other libraries

Blocks that use vanilla JavaScript, jQuery, or other JavaScript frameworks for interactivity are **not compatible** with client-side navigation. When the router replaces a region's content, any event listeners attached via these libraries are lost, and the block's JavaScript will not re-initialize for the new content.

For these blocks, leave `supports.interactivity` at its default value (`false`) or explicitly set `clientNavigation` to `false`:

```json
{
    "supports": {
        "interactivity": {
            "clientNavigation": false
        }
    }
}
```

<div class="callout callout-warning">

> If a block that is not compatible with client-side navigation is detected inside a router region, WordPress may automatically disable client-side navigation on that page to prevent broken behavior. For example, the Query block's enhanced pagination checks for incompatible blocks and falls back to full page reloads when they are found.

Reviewer: This is not exactly true now, as blocks have to do it manually, but it's something we can introduce automatically in the future. 

Maybe we can add a section here explaining how the Query block is doing it, providing an example of how it would be done, and linking to the `clientNavigationDisabled` disabled configuration.

Author: Fixed in https://github.com/WordPress/gutenberg/pull/75263/commits/114abb9cb2418e4af47733f475bcddd879e07a3d.

Use `getServerState()` and `getServerContext()` to react specifically to server-provided values and selectively update the client state in a callback:

```js
import {
    store,
    getContext,
    getServerState,
    getServerContext,
} from '@wordpress/interactivity';

const { state } = store( 'myPlugin', {
    callbacks: {
        syncWithServer() {
            const serverState = getServerState();
            const serverContext = getServerContext();
            const context = getContext();

            // Keep the product count in sync with the server across navigations.
            if ( serverState.productCount !== undefined ) {
                state.productCount = serverState.productCount;
            }

            // Reset the expanded state based on the new page's context.
            if ( serverContext.isExpanded !== undefined ) {
                context.isExpanded = serverContext.isExpanded;
            }
        },
    },
} );
```

> For more details, see the [Understanding global state, local context, and derived state](/docs/reference-guides/interactivity-api/core-concepts/understanding-global-state-local-context-and-derived-state.md#subscribing-to-server-state-and-context) guide.

Reviewer: The name of this file will need to be changed when this pull request is merged.

- https://github.com/WordPress/gutenberg/pull/75357

Author: Fixed in https://github.com/WordPress/gutenberg/pull/75263/commits/3d905755bf07fdbd20fff911e709b90b9bac6eb1.

### Overriding cached pages

> By default, once a page is cached, subsequent navigations use the cached version. Use the `force` option to re-fetch a page even if it's cached:

Reviewer: We should clarify here that this is for overriding the router's internal cache, not the server's cache. This needs to be very clear.

Author: Clarified in https://github.com/WordPress/gutenberg/pull/75263/commits/3b6f0870407420d6d4fde000a297c932acc34d55.

Reviewer: Great, thanks. Maybe we could also change the section title so that it also includes the type of cache we are talking about.

`### Overriding router's internal in-memory cached pages`

            <h1>Custom Content</h1>
            <p>This HTML was provided directly, not fetched.</p>
        </div>
    `,
} );

// Prefetch with custom HTML.
yield actions.prefetch( '/custom-page/', {
    html: customHtmlString,
} );
```

This is useful for:
- Optimistic UI updates where you construct the expected HTML before the server responds.
- Offline scenarios where you provide cached or fallback content.
- Testing and development.

### Managing browser history

By default, `navigate()` adds a new entry to the browser's session history using `pushState`. Use the `replace` option to replace the current history entry instead:

```js
// Default behavior: adds new history entry (pushState).
yield actions.navigate( '/page-2/' );

// Replace current history entry (replaceState).
yield actions.navigate( '/page-2/', { replace: true } );
```

Use `replace: true` when:
- Implementing redirects where the original URL shouldn't be in history.
- Updating query parameters for filtering/sorting where each change shouldn't be a separate history entry.
- Implementing infinite scroll where you update the URL but don't want each page to be a separate history entry.

### Changing the timeout

> Navigation will abort if it takes too long. The default timeout is 10 seconds. Use the `timeout` option to change this:

Reviewer: It would be good to clarify here that if it reaches the timeout, the navigation is not aborted, but rather a fallback to server navigation is made.

This is explained in the next section, but if you only read this section, it gives you the impression that navigation simply doesn't happen because of the use of the word "abort".

Author: Clarified in https://github.com/WordPress/gutenberg/pull/75263/commits/4f3a7b63acdee7a2fbac9d1a899d8d3ca7415e57.

```js
store( 'myPlugin', {
    actions: {
        navigateWithCustomErrorHandling: withSyncEvent( function* ( event ) {
            event.preventDefault();
            const url = event.target.href;

            try {
                // Fetch the page manually.
                const response = yield fetch( url );

                if ( ! response.ok ) {
                    // Handle HTTP errors.
                    state.error = `Error: ${response.status}`;
                    return;
                }

                const html = yield response.text();

                // Navigate using the fetched HTML.
                const { actions } = yield import(
                    '@wordpress/interactivity-router'
                );
                yield actions.navigate( url, { html } );
            } catch ( error ) {
                state.error = 'Network error. Please check your connection.';
            }
        } ),
    },
} );
```

### Disabling client-side navigation on certain pages

Some pages may require a full page reload instead of client-side navigation. Use `wp_interactivity_config()` to disable client navigation:

```php
// In your theme's functions.php or a plugin.

> add_action( 'wp', function() {

Reviewer: I assume you've looked into it, but just doble-checking: is `wp` the correct action to run this?

Author: Yeah, `wp` is supposed to be the earliest reliable hook to call [conditional tags](https://developer.wordpress.org/themes/classic-themes/references/list-of-conditional-tags/) like `is_page()` or `is_page_template()`.

Reviewer: Ok, thanks.

    // Disable on specific page templates.
    if ( is_page_template( 'template-complex.php' ) ) {
        wp_interactivity_config(
            'core/router',
            array( 'clientNavigationDisabled' => true )
        );
    }

    // Disable on admin-like pages.
    if ( is_page( 'dashboard' ) ) {
        wp_interactivity_config(
            'core/router',
            array( 'clientNavigationDisabled' => true )
        );
    }
} );
```

When `clientNavigationDisabled` is `true`:
- `actions.navigate()` triggers a full page reload.
- `actions.prefetch()` does nothing.
- Navigating from another page to this page forces a reload.

### Disabling navigation feedback

The Interactivity API router includes built-in feedback during navigation:

> - **Loading animation**: A progress bar that appears during navigation.

Reviewer: It would be good to specify here what color is used for this bar, which I think is a CSS variable, like `--color-primary` or something like that?

Author: Woops, it looks like the loading bar is not using any CSS variables... 👀 

https://github.com/WordPress/wordpress-develop/blob/45e2aee38fa50cb8a7f60713c390d3f963e17eea/src/wp-includes/interactivity-api/class-wp-interactivity-api.php#L1247

Reviewer: That's interesting. Maybe we had it in the first versions and we forgot to copy it during some migration? I'll put it on the roadmap.

In some cases, you may want to disable these:

```js
// Disable loading animation (for instant-feeling updates).
yield actions.navigate( '/page/', { loadingAnimation: false } );

// Disable screen reader announcements (when providing custom announcements).
yield actions.navigate( '/page/', { screenReaderAnnouncement: false } );

// Disable both.
yield actions.navigate( '/page/', {
    loadingAnimation: false,
    screenReaderAnnouncement: false,
} );
```

Use cases for disabling feedback:
- **Silent updates**: Background refreshes where you don't want to draw attention.
- **Custom loading UI**: When you're implementing your own loading indicators.
- **Custom accessibility**: When you're providing your own screen reader announcements.

### Subscribing to page changes

When using client-side navigation, traditional page load events (like `DOMContentLoaded` or `load`) don't fire because the browser isn't performing a full page reload. If you need to run code after each navigation — for example, to send analytics events or update third-party widgets — you can subscribe to the router's reactive state.

> The `core/router` store exposes a reactive `state.url` property that updates every time a client-side navigation occurs. By reading this value inside a callback, you create a reactive subscription that re-runs whenever the URL changes.

Reviewer: We need a better mechanism for this.

There is also another edge case, which is when you are doing a navigation with `force: true` to the same URL. In that case, the URL doesn't change and the callbacks are not executed again.

For now, maybe it would be good to mention that we will expose a more explicit API in the future.

Author: Done in https://github.com/WordPress/gutenberg/pull/75263/commits/9d1a22f43ac2e773b9f2d6e0be692cec4b738af6.

Author: I guess we'd need to update this section once https://github.com/WordPress/wordpress-develop/pull/10944 is merged. 🤔

Reviewer: Let's include the new (and proper) approach already. I'll do it in a new comment.

```js
// view.js
import { store } from '@wordpress/interactivity';

const routerStore = store( 'core/router' );

// Track the previous URL so we only react to actual navigations.
let previousUrl = window.location.href;

store( 'myAnalytics', {
    callbacks: {
        onPageChange() {
            // Reading state.url creates a reactive subscription.
            const url = routerStore.state.url;

            // Skip if the router hasn't loaded yet or if the URL
            // hasn't actually changed (e.g., initial router load).
            if ( url && url !== previousUrl ) {
                previousUrl = url;
                sendAnalyticsPageView( url );
            }
        },
    },
} );
```

Place this callback on an element that is always present in the DOM (like a site header or a persistent wrapper) using the `data-wp-watch` directive:

```html
<div
    data-wp-interactive="myAnalytics"
    data-wp-watch="callbacks.onPageChange"
>
    <!-- Site content -->
</div>
```

> The router also exposes `state.navigation.hasStarted` and `state.navigation.hasFinished` properties if you need to track the navigation lifecycle more granularly — for example, to show custom loading indicators or to time how long navigations take.

Reviewer: I would delete this since those APIs are internal and, in fact, they shouldn't be part of the official API yet. Actually, I wanted to delete them in [this issue](https://github.com/WordPress/gutenberg/issues/70879).

```suggestion
```

Author: Done in https://github.com/WordPress/gutenberg/pull/75263/commits/6773578e7d7228d5c6b7c7970c1c02cd86fa200e.

                    return;
                }

                const html = yield response.text();

                // Navigate using the fetched HTML.
                const { actions } = yield import(
                    '@wordpress/interactivity-router'
                );
                yield actions.navigate( url, { html } );
            } catch ( error ) {
                state.error = 'Network error. Please check your connection.';
            }
        } ),
    },
} );
```

### Disabling client-side navigation on certain pages

Some pages may require a full page reload instead of client-side navigation. Use `wp_interactivity_config()` to disable client navigation:

```php
// In your theme's functions.php or a plugin.
add_action( 'wp', function() {
    // Disable on specific page templates.
    if ( is_page_template( 'template-complex.php' ) ) {
        wp_interactivity_config(
            'core/router',
            array( 'clientNavigationDisabled' => true )
        );
    }

    // Disable on admin-like pages.
    if ( is_page( 'dashboard' ) ) {
        wp_interactivity_config(
            'core/router',
            array( 'clientNavigationDisabled' => true )
        );

>     }

Reviewer: I had to read it a couple of times until I realized it wasn't talking about the WP admin, but rather some kind of other admin pages generated in the frontend. Since this isn't a common use case in WordPress, I would remove this example to avoid confusion.

Author: Fixed in https://github.com/WordPress/gutenberg/pull/75263/commits/ad3f8e2f6f9840cc57d4eeeb345c3bb85abaee6f.

// Track the previous URL so we only react to actual navigations.
let previousUrl = window.location.href;

store( 'myAnalytics', {
    callbacks: {
        onPageChange() {
            // Reading state.url creates a reactive subscription.
            const url = routerStore.state.url;

            // Skip if the router hasn't loaded yet or if the URL
            // hasn't actually changed (e.g., initial router load).
            if ( url && url !== previousUrl ) {
                previousUrl = url;
                sendAnalyticsPageView( url );
            }
        },
    },
} );
```

Place this callback on an element that is always present in the DOM (like a site header or a persistent wrapper) using the `data-wp-watch` directive:

```html
<div
    data-wp-interactive="myAnalytics"
    data-wp-watch="callbacks.onPageChange"
>
    <!-- Site content -->
</div>
```

<div class="callout callout-info">
The <code>state.url</code> approach has a known limitation: when navigating with <code>force: true</code> to the same URL, the callback won't re-run because the URL value doesn't actually change. A more explicit API for subscribing to navigation events will be provided in the future.

> </div>

Reviewer: What about this?

```suggestion
The `core/router` store exposes a reactive `state.url` property that updates every time a client-side navigation occurs. By reading this value inside a `data-wp-watch` or `watch` callbacks, you create a reactive subscription that re-runs whenever the URL changes.

```js
// view.js
import { watch, store } from '@wordpress/interactivity';

// Store-wide subscription.
watch( () => {
    const { state } = store( 'core/router' );
    sendAnalyticsPageView( state.url );
} );

// Element-based subscription: <div data-wp-watch="callbacks.sendPageView">
store( 'myPlugin', {
    callbacks: {
        sendPageView() {
            const { state } = store( 'core/router' );
            sendAnalyticsPageView( state.url );
            }
        },
    },
} );
```
```

Author: Done in https://github.com/WordPress/gutenberg/pull/75263/commits/820df323449f7e878e0f0d89737c8fb36b87417a.

Once a page is in the cache, it remains there for the duration of the browser session. Subsequent navigations to that URL will use the cached version instantly, without any network request. This is why client-side navigation feels so fast after the initial visit — the page is already prepared and ready to render.

If you need to force a fresh fetch (for example, after submitting a form that changes the page's content), you can use the `force: true` option with `navigate()` or `prefetch()`. This bypasses the cache check and fetches the page anew, replacing the existing cache entry with the fresh content.

### Router regions

Router regions are the sections of your page that the router knows how to update during client-side navigation. They act as boundaries that tell the router "this is the content that should change when navigating between pages."

**Defining router regions**

> You define a router region by adding the `data-wp-router-region` attribute to an element. The element must also be interactive (either having `data-wp-interactive` directly, or being a descendant of an element with `data-wp-interactive`).

Reviewer: We haven't fixed the fact that the `data-wp-router-region` needs a `data-wp-interactive` to inherit the namespace yet, right?

Oh, I see you explain it later. I don't think this part is clear enough.

Author: Clarified in https://github.com/WordPress/gutenberg/commit/103963bbfadbefdc4db687139e85bf5e7c32ce48.

The attribute value serves as a unique identifier for that region. You can specify it in two ways:

1. As a simple string:
   ```html
   <div
       data-wp-interactive="myPlugin"
       data-wp-router-region="myPlugin/main-content"
   >
       <!-- Region content -->
   </div>
   ```

2. As a JSON object (when you need the `attachTo` feature, explained later):
   ```html
   <div
       data-wp-interactive="myPlugin"
       data-wp-router-region='{ "id": "myPlugin/modal", "attachTo": "body" }'
   >
       <!-- Region content -->
   </div>
   ```

The region ID must be unique within a single page and consistent across pages that share the same region. For example, if both your "Products" page and "Product Detail" page have a sidebar, and you want that sidebar to update during navigation, both pages should define a region with the same ID (e.g., `"myPlugin/sidebar"`).

> **Requirements for router regions**

Reviewer: This part shouldn't be in the "In depth" guide, but rather explained the first time `router-regions` are introduced, as it's part of how they work and it's important.

Author: Fixed in https://github.com/WordPress/gutenberg/commit/103963bbfadbefdc4db687139e85bf5e7c32ce48.

First, the router parses the fetched HTML into a document structure using the browser's built-in HTML parser. This gives it a complete DOM tree to work with, just as if the page had been loaded normally.

Next, the router scans this document for all elements that have both `data-wp-interactive` and `data-wp-router-region` attributes. For each region found, it extracts the region ID and checks whether the region is nested inside another region. Only top-level regions are processed directly; nested regions are handled as part of their parent's content.

For each top-level region, the router converts the HTML into a virtual DOM (vDOM) representation. The virtual DOM is a lightweight JavaScript object structure that mirrors the actual DOM but can be compared and manipulated much more efficiently. Importantly, the region element itself is included in this conversion — not just its children. This means that attributes on the region element, such as `data-wp-context`, will also be processed and updated during navigation.

> Finally, each region's virtual DOM is stored in the page cache entry, indexed by its region ID. The cache entry now contains a map of region IDs to their corresponding virtual DOM trees.

Reviewer: Maybe we should also explain here that it also downloads the CSS and JavaScript that are not haven't been downloaded yet?

Author: Mentioned in https://github.com/WordPress/gutenberg/pull/75263/changes/9cc299d9df2ebc5bd8905056f8afbe3628edb944.

```html
<script
    type="module"
    src="/wp-content/plugins/my-plugin/view.js"
    data-wp-router-options='{"loadOnClientNavigation": true}'
></script>
```

When the router fetches a page, it scans for all `<script type="module">` elements that have this attribute with `loadOnClientNavigation` set to `true`. These are the modules it will preload and execute.

**Processing the import map**

Modern JavaScript uses import maps to resolve bare module specifiers (like `@wordpress/interactivity`) to actual URLs. WordPress generates an import map that tells the browser where to find each module:

```html
<script type="importmap">
{
    "imports": {
        "@wordpress/interactivity": "/wp-includes/js/dist/interactivity.min.js",
        "@wordpress/interactivity-router": "/wp-includes/js/dist/interactivity-router.min.js"
    }
}
</script>
```

When the router fetches a new page, it extracts the import map from that page and merges any new mappings with the current page's import map. This ensures that modules can resolve their dependencies correctly even when navigating between pages that have different sets of scripts.

**Preloading modules and their dependencies**

> Preloading script modules is more complex than preloading styles because modules can import other modules. A single entry-point module might depend on dozens of other modules, which might depend on dozens more.

Reviewer: I'm not sure if this sentence makes much sense, as styles can also import other styles, and therefore people will wonder if this works well or not.

Author: Fixed in https://github.com/WordPress/gutenberg/pull/75263/changes/4e078ae62c2d4338183f4c611845650168c987a2.

The Interactivity API provides client-side navigation through the `@wordpress/interactivity-router` package. The central concept is the **router region**: a section of your page that the router knows how to update during navigation. You mark these sections with the `data-wp-router-region` directive, and when the user navigates to a new URL, the router fetches the destination page and replaces only the content inside matching regions — leaving everything else on the page untouched.

This is known as **region-based client-side navigation**, and it is the recommended approach for implementing client-side navigation in WordPress. There is also an experimental **full-page client-side navigation** mode, which treats the entire `<body>` element as a single region — effectively updating the whole page content without a traditional reload. Full-page navigation is covered at the end of this guide in [Full-page client-side navigation (experimental)](#full-page-client-side-navigation-experimental).

> ## Table of contents

Reviewer 2: The Block Editor handbook provides its own table of contents sidebar that remains while scrolling. 
I would avoid adding a custom table of contents.

Author: I'll remove it. Thanks!

Author: Done in https://github.com/WordPress/gutenberg/pull/75263/commits/2b38862be6f08bd8a5f6e06ef005843f0f7581cf.
